webui/react/src/hooks/useSettings.tsx

```diff
@@ -0,0 +1,332 @@
+  if (type === BaseType.Boolean) return param === 'true';
+  if (type === BaseType.Float || type === BaseType.Integer) {
+    const value = Number(param);
+    return !isNaN(value) ? value : undefined;
+  }
+  if (type === BaseType.Object) return JSON.parse(param);
+  if (type === BaseType.String) return param;
+  return undefined;
+};
+
+const queryToSettings = <T, >(config: SettingsConfig<T>, query: string) => {
+  const params = queryString.parse(query);
+
+  return (Object.values(config.settings) as SettingsConfigProp<T>[]).reduce<Settings>((acc, setting) => {
+    /*
+    * Attempt to decode the query parameter and if anything
+    * goes wrong, set it to the default value.
+    */
+    try {
+      const paramValue = params[setting.storageKey];
+      const baseType = setting.type.baseType;
+      const expectingArray = setting.type.isArray;
+
+      if (paramValue !== null) {
+        let queryValue: Primitive | Primitive[] | undefined;
+        /*
+        * Convert the string-based query params to primitives.
+        * `undefined` values can happen if the query param values are invalid.
+        *   string[] => Primitive[] | undefined
+        *   string   => Primitive | undefined
+        */
+        if (Array.isArray(paramValue)) {
+          queryValue = paramValue
+            .reduce<Primitive[]>((acc, value) => {
+              const parsedValue = queryParamToType(baseType.name as BaseType, value);
+
+              if (parsedValue !== undefined) acc.push(parsedValue);
+
+              return acc;
+            }, []);
+        } else {
+          queryValue = queryParamToType(baseType.name as BaseType, paramValue);
+        }
+
+        if (queryValue !== undefined) {
+          /*
+          * When expecting an array, convert valid non-array values into an array.
+          * Example - 'PULLING' => [ 'PULLING' ]
+          */
+          const normalizedValue = (() => {
+            if (expectingArray && !Array.isArray(queryValue)) {
+              return [queryValue];
+            }
+            return queryValue;
+          })();
+
+          acc[setting.storageKey] = normalizedValue;
+        }
+      }
+    } catch (e) {
+      handleError(e, { silent: true, type: ErrorType.Ui });
+    }
+
+    return acc;
+  }, {});
+};
+
+const useSettings = <T, >(config: SettingsConfig<T>): UseSettingsReturn => {
+  const { auth: { user } } = useStore();
+  const { state, update } = useContext(UserSettings);
+  const navigate = useNavigate();
+
+  // creates an entry at the global state
+  useEffect(() => {
+    const settings = state.get(config.applicableRoutespace);
+
+    if (settings) return;
+
+    const newSettings: Settings = {};
+
+    for (const key in config.settings) {
+      const setting = config.settings[key];
+
+      newSettings[setting.storageKey] = setting.defaultValue;
+    };
+
+    state.set(config.applicableRoutespace, newSettings);
+  }, [config, state]);
+
+  // parse navigation url to state
+  useEffect(() => {
```

Comment:
What do you think about putting this logic in the `SettingsProvider` instead of the hook? It only needs to happen once on page load since we only ever want to restore settings based on the initial navigation. Here this might cause unintended state manipulations.

Reply:
I don't think that's the intended behavior. This logic is meant to handle when the user navigates directly to the page. When that happens we want to restore settings–the same way we restore them from the DB when not navigating directly. When you're navigating _within_ the app, we don't need to read the url parameters because the _app_ is setting the url.

Are there cases where we're using `navigate` to bounce settings through the url instead of setting them directly?